pkg/pgp/key.go

```diff
@@ -59,7 +60,15 @@ func (p *Key) Verify(data, signature []byte) error {
 
 	sig := pgpcrypto.NewPGPSignature(signature)
 
-	return p.keyring.VerifyDetached(message, sig, pgpcrypto.GetUnixTime())
+	clockSkew := DefaultAllowedClockSkew.Seconds()
+
+	i := p.key.GetEntity().PrimaryIdentity()
+
+	if i.SelfSignature.KeyLifetimeSecs != nil {
+		clockSkew = math.Min(float64(*i.SelfSignature.KeyLifetimeSecs)/2, clockSkew)
```

Comment:
should we have a test for that? can we?

does this actually shorten the validitiy of the key? if we set current time in the future?

Reply:
I couldn't find a way to mock time skew in the tests...

> does this actually shorten the validitiy of the key? if we set current time in the future?

Yeah... Another option is to do 2 attempts, one without skew, another with the skew.

Reply:
I think we can probably mock the time if we need to for test purposes, or at least have a way to update it, and verify that this fix doesn't break things. @DmitriyMV might have some context on that?

Reply:
Added a test that at least validates upper boundary, when the key expires.
Seems to work fine even without any changes. I guess that's due to that the library was already handling skew to the past properly.

Reply:
> Another option is to do 2 attempts, one without skew, another with the skew.

I think it already does this [here](https://github.com/ProtonMail/gopenpgp/blob/b97d994962e4380235fc1bff8a599eacf6a120ff/crypto/signature.go#L243) and [here](https://github.com/ProtonMail/gopenpgp/blob/b97d994962e4380235fc1bff8a599eacf6a120ff/crypto/signature.go#L258).

I do not understand what test below does.

Reply:
As for mocking - you could probably play with `github.com/ProtonMail/gopenpgp/v2/crypto.UpdateTime`

Reply:
> I think it already does this [here](https://github.com/ProtonMail/gopenpgp/blob/b97d994962e4380235fc1bff8a599eacf6a120ff/crypto/signature.go#L243) and [here](https://github.com/ProtonMail/gopenpgp/blob/b97d994962e4380235fc1bff8a599eacf6a120ff/crypto/signature.go#L258).

I've debugged this code and it seems weird to me: it shifts the time by 2 days (!) in the future and never hits our key validity interval.

Test creates a key valid for 1 second and tries to use it several times, allowed time skew there is 0.5s, so it works 2 times: in the beginning of the valid period and in the end.

Reply:
> github.com/ProtonMail/gopenpgp/v2/crypto.UpdateTime

Works for the upper boundary check, but can't do checks for issued in the future, do to this if:
https://github.com/ProtonMail/gopenpgp/blob/b97d994962e4380235fc1bff8a599eacf6a120ff/crypto/time.go#L12
Can't rewind the time back.

Reply:
Actually hold on, I think I can actually do it but only once.